pkg/cmd/webhooks/forward.go

```diff
@@ -0,0 +1,253 @@
+				defer close(done)
+				conn, err := handleWebsocket(token, wsURL, &opts, events)
+
+				if err != nil {
+					unwrapped := errors.Unwrap(err)
+					var syscallErr *os.SyscallError
+					// If the error is a TCP handleWebsocket error, retry connecting
+					if errors.As(unwrapped, &syscallErr) && (unwrapped.Error() == "connect: connection refused") {
+						retries += 1
+						time.Sleep(5 * time.Second)
+						continue
+					}
+					// If the error is a server disconnect (1006), retry connecting
+					if websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+						retries += 1
+						time.Sleep(10 * time.Second)
+						continue
+					}
+
+					if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
+						return nil
+					}
+					return err
+				}
+
+				fmt.Fprintf(opts.IO.Out, "Connected to webhooks server, forwarding \n")
+				retries = 0
+
+				go func() {
+					for {
+						select {
+						case <-done:
+							return
+						case reply := <-replies:
+							err := conn.WriteJSON(reply)
+							if err != nil {
+								return
+							}
+						}
+					}
+				}()
+			}
+			return nil
+		},
+	}
+	cmd.Flags().StringVarP(&opts.EventType, "event", "E", "", "Name of the event type to forward")
+	cmd.Flags().StringVarP(&opts.Repo, "repo", "R", "", "Name of the repo where the webhook is installed")
+	cmd.Flags().IntVarP(&opts.Port, "port", "P", 0, "Local port to receive webhooks on")
+	cmd.Flags().StringVarP(&opts.Host, "host", "H", "", "Host address of GitHub API")
+	return &cmd
+}
+
+// createHook issues a request against the GitHub API to create a dev webhook
+func createHook(o *hookOptions) (string, error) {
+	httpClient, err := o.HttpClient()
+	if err != nil {
+		return "", err
+	}
+	apiClient := api.NewClientFromHTTP(httpClient)
+	path := fmt.Sprintf("repos/%s/hooks", o.Repo)
+	req := createHookRequest{
+		Name:   "cli",
+		Events: []string{o.EventType},
+		Active: true,
+		Config: hookConfig{
+			ContentType: "json",
+			InsecureSSL: "0",
+		},
+	}
+
+	reqBytes, err := json.Marshal(req)
+	if err != nil {
+		return "", err
+	}
+	var res createHookResponse
+	err = apiClient.REST(o.Host, "POST", path, bytes.NewReader(reqBytes), &res)
+	if err != nil {
+		return "", err
+	}
+	return res.WsURL, nil
+}
+
+// handleWebsocket connects to the websocket server and saves received events to a channel
+func handleWebsocket(token string, url *url.URL, opts *hookOptions, events chan wsEventReceived) (*websocket.Conn, error) {
+	fmt.Fprintf(opts.IO.Out, "Attempting to connect to webhooks server...\n")
+	h := make(http.Header)
+	h.Set("Authorization", token)
+
+	c, resp, err := websocket.DefaultDialer.Dial(url.String(), h)
+	if err != nil {
+		if resp != nil {
+			bts, _ := io.ReadAll(resp.Body)
+			err = fmt.Errorf("ws err %d - %s - %v", resp.StatusCode, bts, err)
+		}
+		return nil, err
+	}
+
+	defer c.Close()
+
+	for {
+		var ev wsEventReceived
+		err := c.ReadJSON(&ev)
+		if err != nil {
+			return nil, err
+		}
+		events <- ev
+	}
```

Comment:
Sine this is an infinite loop and only returns when there's an error, then it will never return the websocket.Connection back to the caller, and therefore will never write the responses back to the websocket server to forwarded back to hookshot-go. 

Which makes me think: I think the `WaitGroup` in your test should probably be on the websocket server side, and not on the handler: In other words, we test that the webhook not only made it all the way to the customer's local server, but also all the way back to the websocket server.

Reply:
🤦🏻‍♀️ Somehow I forgot that the replies have to make it back to the ws server, and eventually to the UI. I refactored the code, now we're sending the response back. And it's also easier to follow, I think.